From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] ole32: Marshal FLAG_STGMEDIUM and IDataObject::SetData.

FLAG_STGMEDIUM is marshalled (ContextFlags, fPassOwnership, then the
STGMEDIUM; freed unless ownership passed), and IDataObject::SetData's
call_as proxy and stub use it, so SetData works through a proxy. All were
no-ops returning E_NOTIMPL.

Gate: test/ole32-setdata-gate.sh (64/32).
---
 dlls/ole32/usrmarshal.c | 55 ++++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 45 insertions(+), 10 deletions(-)

diff --git a/dlls/ole32/usrmarshal.c b/dlls/ole32/usrmarshal.c
index f3f0b32..bdd06cd 100644
--- a/dlls/ole32/usrmarshal.c
+++ b/dlls/ole32/usrmarshal.c
@@ -1111,27 +1111,51 @@ void __RPC_USER ASYNC_STGMEDIUM_UserFree(ULONG *pFlags, ASYNC_STGMEDIUM *pStgMed
     STGMEDIUM_UserFree(pFlags, pStgMedium);
 }
 
+/* FLAG_STGMEDIUM: ContextFlags and fPassOwnership, then the STGMEDIUM
+ * (userFLAG_STGMEDIUM in objidl.idl) */
 ULONG __RPC_USER FLAG_STGMEDIUM_UserSize(ULONG *pFlags, ULONG StartingSize, FLAG_STGMEDIUM *pStgMedium)
 {
-    FIXME(":stub\n");
-    return StartingSize;
+    ULONG size = StartingSize;
+
+    TRACE("%s, %lu, %p.\n", debugstr_user_flags(pFlags), StartingSize, pStgMedium);
+
+    ALIGN_LENGTH(size, 3);
+    size += 2 * sizeof(DWORD);
+    return STGMEDIUM_UserSize(pFlags, size, &pStgMedium->Stgmed);
 }
 
 unsigned char * __RPC_USER FLAG_STGMEDIUM_UserMarshal(  ULONG *pFlags, unsigned char *pBuffer, FLAG_STGMEDIUM *pStgMedium)
 {
-    FIXME(":stub\n");
-    return pBuffer;
+    TRACE("%s, %p, %p.\n", debugstr_user_flags(pFlags), pBuffer, pStgMedium);
+
+    ALIGN_POINTER(pBuffer, 3);
+    *(DWORD *)pBuffer = pStgMedium->ContextFlags;
+    pBuffer += sizeof(DWORD);
+    *(DWORD *)pBuffer = pStgMedium->fPassOwnership;
+    pBuffer += sizeof(DWORD);
+    return STGMEDIUM_UserMarshal(pFlags, pBuffer, &pStgMedium->Stgmed);
 }
 
 unsigned char * __RPC_USER FLAG_STGMEDIUM_UserUnmarshal(ULONG *pFlags, unsigned char *pBuffer, FLAG_STGMEDIUM *pStgMedium)
 {
-    FIXME(":stub\n");
-    return pBuffer;
+    TRACE("%s, %p, %p.\n", debugstr_user_flags(pFlags), pBuffer, pStgMedium);
+
+    ALIGN_POINTER(pBuffer, 3);
+    pStgMedium->ContextFlags = *(DWORD *)pBuffer;
+    pBuffer += sizeof(DWORD);
+    pStgMedium->fPassOwnership = *(DWORD *)pBuffer;
+    pBuffer += sizeof(DWORD);
+    memset(&pStgMedium->Stgmed, 0, sizeof(pStgMedium->Stgmed));
+    return STGMEDIUM_UserUnmarshal(pFlags, pBuffer, &pStgMedium->Stgmed);
 }
 
 void __RPC_USER FLAG_STGMEDIUM_UserFree(ULONG *pFlags, FLAG_STGMEDIUM *pStgMedium)
 {
-    FIXME(":stub\n");
+    TRACE("%s, %p.\n", debugstr_user_flags(pFlags), pStgMedium);
+
+    /* a medium passed with ownership belongs to the callee now */
+    if (!pStgMedium->fPassOwnership)
+        STGMEDIUM_UserFree(pFlags, &pStgMedium->Stgmed);
 }
 
 ULONG __RPC_USER SNB_UserSize(ULONG *pFlags, ULONG StartingSize, SNB *pSnb)
@@ -2030,8 +2054,19 @@ HRESULT CALLBACK IDataObject_SetData_Proxy(
     STGMEDIUM *pmedium,
     BOOL fRelease)
 {
-    FIXME(":stub\n");
+    FLAG_STGMEDIUM flag_medium;
+
+    TRACE("%p, %p, %p, %d.\n", This, pformatetc, pmedium, fRelease);
+
+    /* the medium is [unique]: NULL goes to the object as it is */
+    flag_medium.ContextFlags = 0;
+    flag_medium.fPassOwnership = fRelease;
+    if (pmedium) flag_medium.Stgmed = *pmedium;
+#ifndef SG_MUTANT_SETDATA_NOTIMPL
+    return IDataObject_RemoteSetData_Proxy(This, pformatetc, pmedium ? &flag_medium : NULL, fRelease);
+#else
     return E_NOTIMPL;
+#endif
 }
 
 HRESULT __RPC_STUB IDataObject_SetData_Stub(
@@ -2040,8 +2075,8 @@ HRESULT __RPC_STUB IDataObject_SetData_Stub(
     FLAG_STGMEDIUM *pmedium,
     BOOL fRelease)
 {
-    FIXME(":stub\n");
-    return E_NOTIMPL;
+    TRACE("%p, %p, %p, %d.\n", This, pformatetc, pmedium, fRelease);
+    return IDataObject_SetData(This, pformatetc, pmedium ? &pmedium->Stgmed : NULL, fRelease);
 }
 
 /* call_as/local stubs for oleidl.idl */
